x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
+		tokenOut, err := swapModule.CalcOutAmtGivenIn(
+			ctx, poolI, currFromCoin, req.ToCoinDenom, sdk.ZeroDec(),
+		)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		if tokenOut.IsZero() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
+
+		currTradePrice := sdk.NewDec(currFromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+		priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
+
+		if priceDeviation.LTE(adjustedMaxPriceImpact) {
+			lowAmount = midAmount.Add(sdk.OneInt())
+		} else {
+			highAmount = midAmount.Sub(sdk.OneInt())
+		}
+	}
+
+	// highAmount is 0 it means the loop has iterated to the end without finding a viable trade that respects
+	// the price impact.
+	if highAmount.IsZero() {
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+			OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+		}, nil
+	}
+
+	tokenOut, err = swapModule.CalcOutAmtGivenIn(
+		ctx, poolI, currFromCoin, req.ToCoinDenom, poolI.GetSpreadFactor(ctx),
+	)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+		InputCoin:  currFromCoin,
+		OutputCoin: tokenOut,
+	}, nil
+}
+
+// estimateTradeBasedOnPriceImpactStableSwapPool estimates a trade based on price impact for a stableswap pool type.
+// For a stableswap pool if an amount entered is greater than the total pool liquidity the trade estimated would
+// `panic`. If the amount is small it would return an error, in the case of a `panic` we should ignore it
+// and keep attempting lower input amounts while if it's a normal error we should return an empty trade.
+func (q Querier) estimateTradeBasedOnPriceImpactStableSwapPool(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+	spotPrice, adjustedMaxPriceImpact sdk.Dec,
+	swapModule types.PoolModuleI,
+	poolI types.PoolI,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	var tokenOut sdk.Coin
+	var err error
+	err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+		tokenOut, err = swapModule.CalcOutAmtGivenIn(ctx, poolI, req.FromCoin, req.ToCoinDenom, sdk.ZeroDec())
+		return err
+	})
+
+	// Find out if the error is because the amount is too large or too little. The calculation should error
+	// if the amount is too small, and it should panic if the amount is too large. If the amount is too large
+	// we want to continue to iterate to find attempt to find a smaller value.
+	if err != nil && !strings.Contains(err.Error(), "panic") {
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+			OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
```

Comment:
What's the reason we catch panic here, not in Balancer?

Reply:
Balancer Pool does not panic, while StableSwap does because of the maths involved.

Reply:
Oh wow did not know this, can you add this in the comment please?